core/src/main/java/com/datastax/oss/driver/api/core/cql/Statement.java

```diff
@@ -302,6 +302,50 @@ default SelfT setRoutingKey(@NonNull ByteBuffer... newRoutingKeyComponents) {
   @CheckReturnValue
   SelfT setPagingState(@Nullable ByteBuffer newPagingState);
 
+  /**
+   * Sets the paging state to send with the statement, or {@code null} if this statement has no
+   * paging state.
+   *
+   * <p>This variant uses the "safe" paging state wrapper, it will throw immediately if the
+   * statement doesn't match the one that the state was initially extracted from (same query string,
+   * same parameters). The advantage is that it fails fast, instead of waiting for an error response
+   * from the server.
+   *
+   * <p>Note that, if this statement is a {@link SimpleStatement} with bound values, those values
+   * must be encoded in order to perform the check. This method uses the default codec registry and
+   * default protocol version. This might fail if you use custom codecs; in that case, use {@link
+   * #setPagingState(PagingState, Session)} instead.
+   *
+   * @throws IllegalArgumentException if the given state does not match this statement.
+   * @see #setPagingState(ByteBuffer)
+   * @see ExecutionInfo#getSafePagingState()
+   */
+  @NonNull
+  @CheckReturnValue
+  default SelfT setPagingState(@Nullable PagingState newPagingState) {
+    return setPagingState(newPagingState, null);
+  }
+
+  /**
+   * Alternative to {@link #setPagingState(PagingState)} that specifies the session the statement
+   * will be executed with. <b>You only need this for simple statements, and if you use custom
+   * codecs.</b> Bound statements already know which session they are attached to.
+   */
+  @NonNull
+  @CheckReturnValue
+  default SelfT setPagingState(@Nullable PagingState newPagingState, @Nullable Session session) {
+    if (newPagingState == null) {
+      return setPagingState((ByteBuffer) null);
+    } else if (newPagingState.matches(this, session)) {
+      return setPagingState(newPagingState.getRawPagingState());
+    } else {
+      throw new IllegalArgumentException(
+          "Paging state mismatch, "
```

Comment:
Is it possible to give more info about what's wrong with the non-matched paging state for the debug purpose?
(I.e. which variable is missing)

Reply:
That information is not available, the hash is all we have left about the old statement.